<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Reproducible research in R</title>
    <meta charset="utf-8" />
    <meta name="author" content="Roy Francis and Mun-Gwan Hong" />
    <meta name="keywords" content="r, RaukR, markdown" />
    <script src="libs/header-attrs-2.14/header-attrs.js"></script>
    <link href="libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
    <link href="libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
    <link id="font-awesome-1-attachment" rel="attachment" href="libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>
    <script src="libs/kePrint-0.0.1/kePrint.js"></script>
    <link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
    <link rel="stylesheet" href="assets/presentation.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Reproducible research in R
]
.subtitle[
## RaukR 2022 • Advanced R for Bioinformatics
]
.author[
### <b>Roy Francis and Mun-Gwan Hong</b>
]
.institute[
### NBIS, SciLifeLab
]

---

exclude: true
count: false


&lt;link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;amp;subset=latin-ext" rel="stylesheet"&gt;
&lt;link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.3.1/css/all.css" integrity="sha384-mzrmE5qonljUremFsqc01SB46JvROS7bZs3IO2EmfFsd15uHvIt+Y8vEf7N7fWAU" crossorigin="anonymous"&gt;





&lt;!-- ----------------- Only edit title &amp; author above this ----------------- --&gt;


---

name: topics

## Topics

* Reproducibility
* Environment
* RStudio
* Markdown/Rmarkdown
* Reports and presentations in R

---
name: fuss

## What's all the fuss about?

&lt;img src="rr_presentation_assets/nature-reproducibility.jpg" class="fancyimage size-70"&gt;  
.small[&lt;https://www.nature.com/collections/prbfkwmwvz/&gt;]  

&lt;img src="rr_presentation_assets/nature-reproducibility-2.jpg" class="fancyimage size-70"&gt;  
.small[&lt;https://www.nature.com/news/1-500-scientists-lift-the-lid-on-reproducibility-1.19970&gt;]  

???

A large percentage of research is not reproducible by other researchers or by the original researchers themselves. This concern has been lately addressed by several high profile journals.

---

&lt;img src="rr_presentation_assets/nature-rr-pie.jpg" class="fancyimage size-45"&gt;  
&lt;img src="rr_presentation_assets/nature-rr-bar.jpg" class="fancyimage size-65"&gt;  

---

## What is reproducibility?

&gt; "reproducibility refers to the **ability** of a researcher **to duplicate the results** of a prior study using the same materials as were used by the original investigator. That is, a second researcher might use the same raw data to build the same analysis files and implement the same statistical analysis in an attempt to yield the same results…"

.small[K. Bollen, J. T. Cacioppo, R. Kaplan, J. Krosnick, J. L. Olds, Social, Behavioral, and Economic Sciences Perspectives on Robust and Reliable Science (National Science Foundation, Arlington, VA, 2015)]

.small[Goodman, S. N., Fanelli, D., &amp; Ioannidis, J. P. A. (2016). What does research reproducibility mean? Science Translational Medicine, 8(341), 341ps12–341ps12. http://doi.org/10.1126/scitranslmed.aaf5027]

---

## Reproducibility in R

&lt;img src="rr_presentation_assets/turingway-reproduciblematrix.jpeg" class="size-85"&gt;  

.small[from Turing way (&lt;https://the-turing-way.netlify.app/reproducible-research/overview/overview-definitions.html&gt;)]

---
class: spaced

## Typical workflow

1. Get data
2. Clean, transform data in spreadsheet
3. Copy-paste, copy-paste, ...
4. Run analysis &amp; make figures
5. Write up a report attaching the figures
6. Need to modify the source data/analysis
7. Go back to step 2, Repeat
8. (after a couple of months) Need to fix the figures
9. Ask to myself which data I'm supposed to use now? (`data-final3` or `data-refined2`)
10. Realize the number of samples didn't match
11. Back to step 2, try to remember why some data was modified manually before

--

* (a couple of years later) "I got these figures from you. Could you modify them?"
.size-30[![](rr_presentation_assets/picard.jpg)]

???

Manually handling workflow is hard to reproduce because it is hard to know the exact step carried out. A programmatic workflow allows full transparency to the exact steps followed.

---
class: spaced

## Benefits of reproducibility


- Get the same results as before
- Rerunning workflow
- Additional data/New data
- Returning to a project
- Transferring projects
- Collaborative work
- Easy to make changes
- Eliminate copy-paste errors

???

A reproducible workflow allows a lot of convenience. 

- It's easy to automate re-running of analysis when earlier steps have changed such as new input data, code or assumptions. 
- Useful for an investigator returning to an analyses after a period of time.
- Useful when a project is transferred to a new investigator.
- Useful when working collaboratively.
- When you are asked to modify or change a parameter.

---
class: spaced

## Solutions

![](rr_presentation_assets/rr-solutions.jpg)

* Containerised computing environment - *eg: Singularity, Docker*
* Workflow manager - *eg: Snakemake, Nextflow*
* Package and environment manager - *eg: renv, Conda*
* Track edits and collaborate coding - *eg: Git*
* Share and track code - *eg: GitHub, Bitbucket*
* Notebooks to document ongoing analyses - *eg: Jupyter*
* Analyze and generate reports - *eg: R Markdown*

A NBIS course of [**Tools for reproducible research**](https://nbis-reproducible-research.readthedocs.io/en/latest/)

???

Reproducible projects can be performed at different levels. Reproducibility is the ability for a work to be reproduced by an independently working third-party.

---
class: spaced

## Steps to reproducibility

* Avoid manual steps
* Documents containing analysis, code and results
* Note the software environment
* Self-contained portable project
* Results directly linked to code
* Contextual narrative to why a certain step was performed
* Version control of documents
* Keep the original data intact (read-only) with descriptions incl. how the data was obtained

???

Reproducible programming is not an R specific issue. R offers a set of tools and an environment that is conducive to reproducible research.

---

## Environment

* Not about `environment()`
* Software environment around your code
* Operating system (Windows, Mac, Linux, ...)
* A particular version of R/Python
* Loaded package versions

---

## Software for environment management

* Operating system - *Singularity*, *Docker*

* R/Python - *Conda*

* Loaded package versions - *`renv`* package

---

## `renv` package

* **R env**ironment management package

* It helps our individual projects **isolated** so **portable**, and **reproducible**

* Local library of R packages

* Install `renv` package from CRAN or Conda

```
install.packages("renv")
conda install -c conda-forge r-renv 
```

* Initialize local R environment using `renv`

```
renv::init()
```  

* Save the local library state

```
renv::snapshot()
```

* Restore the local library

```
renv::restore()
```

.small[&lt;https://kevinushey-2020-rstudio-conf.netlify.app/slides.html&gt;]

---

## Install R packages

* Use `renv::install` as below
* From CRAN : `renv::install("`*package name*`")` (e.g. `renv::install("dplyr")`)
* From Bioconductor : `renv::install("bioc::`*package name*`")` 
(e.g `renv::install("bioc::Biobase")`)
* From GitHub : `renv::install("`*user name*`/`*repository*`")` (e.g. `renv::install("StoreyLab/qvalue")`) 
* From GitLab/Bitbucket : `renv::install("[gitlab|bitbucket]::`*user name*`/`*repository*`")`

---

## RStudio • IDE

&lt;img src="rr_presentation_assets/rstudio.jpg" class="fancyimage size-90"&gt;

* Code completion &amp; Syntax highlighting (for many languages)
* R Notebook
* Debugging
* Useful GUI elements
* Multiple sessions can be opened in parallel

---

## RStudio • Project

.small[**Create a new project**]
&lt;img src="rr_presentation_assets/new-project.gif" class="fancyimage size-90"&gt;

* Portable project (.Rproj)
* Dynamic reports
* Version control (git)
* Package management (`renv`)

---

## Project Structure

```
project_name/
+-- data/
|   +-- gene_counts.txt
|   +-- metadata.txt
+-- results/
|   +-- gene_filtered_counts.txt
|   +-- gene_vst_counts.txt
+-- images/
|   +-- exp-setup.jpg
+-- scripts/
|   +-- bash/
|   |   +-- fastqc.sh
|   |   +-- trim_adapters.sh
|   |   +-- mapping.sh
|   +-- r/
|       +-- qc.R
|       +-- functions.R
|       +-- dge.R
+-- report/
    +-- report.Rmd
```

* Organise data, scripts and results sensibly
* Keep projects self contained
* Use relative links

???

Try to organize all material related to a project in a common directory. Organise the directory in a sensible manner. Use relative links to refer to files. Consider raw as read-only content.

---

## Document converter

![](rr_presentation_assets/knit.png)

- Rmd &gt; md &gt; docx|HTML|PDF
- PDF needs Latex
- Handouts
- Scientific Articles
- Presentations
  - beamer
  - ioslides
  - slidy
  - xaringan

---

## Document formats

- [Summary](https://rmarkdown.rstudio.com/formats.html)
- Reports in [HTML](https://bookdown.org/yihui/rmarkdown/html-document.html), [PDF](https://bookdown.org/yihui/rmarkdown/pdf-document.html), [MS Word](https://bookdown.org/yihui/rmarkdown/word-document.html) etc
- Simple web pages and websites using [Rmarkdown](https://rmarkdown.rstudio.com/lesson-13.html)
- Complex websites using [blogdown](https://bookdown.org/yihui/blogdown/)
- Books using [bookdown](https://bookdown.org/yihui/bookdown/)
- Package documentation using [pkgdown](http://pkgdown.r-lib.org/)
- Web applications and interactive documents using [Shiny](https://www.rstudio.com/products/shiny/)
- Dashboards using [flexdashboard](https://rmarkdown.rstudio.com/flexdashboard/) or [shinydashboard](https://rstudio.github.io/shinydashboard/)

---

## RMarkdown • Intro

### Markdown

- Plain text format for readability
- Support pure HTML for complex formatting
- Many variations
  - [John Gruber's original](https://daringfireball.net/projects/markdown/syntax)
  - [GitHub Flavored Markdown (GFM)](https://github.github.com/gfm/)
  - [Pandoc](https://pandoc.org/MANUAL.html#pandocs-markdown)
  - [MultiMarkdown](https://fletcherpenney.net/multimarkdown/)
  - [![](rr_presentation_assets/cm.png)](https://commonmark.org/)
- Pandoc supports conversion to multiple output formats
- To compare MD variants [![](rr_presentation_assets/bm.png)](https://babelmark.github.io)

--

### RMarkdown

- Markdown + embedded R chunks
- Combine text and code in one file
- RMarkdown mostly uses [Pandoc markdown](https://rmarkdown.rstudio.com/authoring_pandoc_markdown.html%23raw-tex#pandoc_markdown)

---

## RStudio • Notebook

.small[**Create a new .Rmd document**]
&lt;img src="rr_presentation_assets/new-rmarkdown.gif" class="fancyimage"&gt;

* Text and code can be written together
* Inline R output (text and figures)

???

R Notebook demonstration.

---

## RMarkdown • Guide

* Create a file that ends in `.Rmd`
* Add YAML matter to top

```
---
title: "This is a title"
output: 
  rmarkdown::html_document
---
```
* In RStudio `File &gt; New File &gt; R Markdown` opens up an Rmd template
* Render interactively using the **Knit** button .fancyimage[![](rr_presentation_assets/knit-button.png)]
* Render using command `rmarkdown::render("report.Rmd")`

---

## RMarkdown • Guide

.pull-left-50[
```
### Heading 3


#### Heading 4


_italic text_    *italic text* 
__bold text__    **bold text**
`code text`  
~~strikethrough~~  

2^10^  2~10~  
$2^{10}$  $2_{10}$  

$\sum\limits_{n=1}^{10} \frac{3}{2}\cdot n$


- bullet point

Link to [this](somewhere.com)

![](https://www.r-project.org/Rlogo.png)

```
]

.pull-right-50[

### Heading 3
#### Heading 4

*italic text*  
**bold text**  
`code text`  
~~strikethrough~~  
2&lt;sup&gt;10&lt;/sup&gt;  2&lt;sub&gt;10&lt;/sub&gt;  
`\(2^{10}\)`  `\(2_{10}\)`  
`\(\sum\limits_{n=1}^{10} \frac{3}{2}\cdot n\)`

* bullet point

Link to [this](somewhere.com)

.size-60[![](https://www.r-project.org/Rlogo.png)]
]

---

## RMarkdown • Guide

* R code can be executed inline

Today's date is `` `r date()` ``    
Today's date is Fri Jun 10 11:24:02 2022  


* R code can be executed in code chunks


````{.r}
```{r}
date()
```
````

* By default shows input code and output result.


```r
date()
```

```
## [1] "Fri Jun 10 11:24:02 2022"
```

* Many arguments to [customise chunks](https://yihui.name/knitr/options/)
  * Set `eval=FALSE` to not evaluate a code chunk
  * Set `echo=FALSE` to hide input code
  * Set `results="hide"` to hide output

* [R Markdown reference](https://rmarkdown.rstudio.com/lesson-1.html)

---

## A few packages useful for R Markdown

* [`kableExtra`](https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.html) : Beautiful tables in html

&lt;table class=" lightable-classic table" style='font-family: "Arial Narrow", "Source Sans Pro", sans-serif; margin-left: auto; margin-right: auto; font-size: 10px; width: auto !important; margin-left: auto; margin-right: auto;'&gt;
 &lt;thead&gt;
&lt;tr&gt;
&lt;th style="empty-cells: hide;" colspan="1"&gt;&lt;/th&gt;
&lt;th style="padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="2"&gt;&lt;div style="border-bottom: 1px solid #111111; margin-bottom: -1px; "&gt;Group 1&lt;/div&gt;&lt;/th&gt;
&lt;th style="padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="2"&gt;&lt;div style="border-bottom: 1px solid #111111; margin-bottom: -1px; "&gt;Group 2&lt;/div&gt;&lt;/th&gt;
&lt;th style="padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="2"&gt;&lt;div style="border-bottom: 1px solid #111111; margin-bottom: -1px; "&gt;Group 3&lt;/div&gt;&lt;/th&gt;
&lt;/tr&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt;   &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; mpg &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; cyl &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; disp &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; hp &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; drat &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; wt &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Mazda RX4 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 21.0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 6 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 160 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 110 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 3.90 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 2.620 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Mazda RX4 Wag &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 21.0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 6 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 160 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 110 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 3.90 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 2.875 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Datsun 710 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 22.8 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 4 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 108 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 93 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 3.85 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 2.320 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Hornet 4 Drive &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 21.4 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 6 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 258 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 110 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 3.08 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 3.215 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; Hornet Sportabout &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 18.7 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 8 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 360 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 175 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 3.15 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 3.440 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

* [`english`](https://cran.r-project.org/web/packages/english/vignettes/the-english-patient.pdf) : Integer in English (e.g. `two hundred one` instead of `201`)

* [`janitor`](https://cran.r-project.org/web/packages/janitor/vignettes/janitor.html) : More than `table`


```r
janitor::tabyl(mtcars, cyl) %&gt;% 
  janitor::adorn_pct_formatting()
```

```
##  cyl  n percent
##    4 11   34.4%
##    6  7   21.9%
##    8 14   43.8%
```

---

## RStudio • Project with Git

.small[**Create a new project with version control**]
&lt;img src="rr_presentation_assets/rstudio_git_new.gif" class="fancyimage"&gt;

* Version control : keep old versions and who/when files modified for what
* A repository in GitHub/Bitbucket

---

## RStudio • Git commit

.small[**Log a set of changes using Git**]
&lt;img src="rr_presentation_assets/rstudio_git_push.gif" class="fancyimage"&gt;


???

How to Git commit using RStudio

---

name: help
class: spaced

## Acknowledgements

* [**Reproducible Research in R and RStudio**](https://www.slideshare.net/SusanJohnston3/reproducible-research-in-r-and-r-studio) - Susan Johnston  
* [**New Tools for Reproducible Research with R**](https://slides.yihui.name/2012-knitr-RStudio.html) - JJ Allaire and Yihui Xie  
* [**Reproducible research with R**](http://www.hafro.is/~einarhj/education/tcrenv2016/pre/r-markdown.pdf) - Bjarki Thor Elvarsson and Einar Hjorleifsson  
* [**Reproducible Research Workshop**](http://www.geo.uzh.ch/microsite/reproducible_research/post/rr-r-publication/) - University of Zurich  
* RStudio [learning](https://www.rstudio.com/online-learning/)


&lt;!-- --------------------- Do not edit this and below --------------------- --&gt;

---
name: end-slide
class: end-slide, middle
count: false

# Thank you. Questions?


&lt;p&gt;R version 4.1.3 (2022-03-10)&lt;br&gt;&lt;p&gt;Platform: x86_64-apple-darwin13.4.0 (64-bit)&lt;/p&gt;&lt;p&gt;OS: macOS Big Sur/Monterey 10.16&lt;/p&gt;&lt;br&gt;

Built on : &lt;i class='fa fa-calendar' aria-hidden='true'&gt;&lt;/i&gt; 10-Jun-2022 at &lt;i class='fa fa-clock-o' aria-hidden='true'&gt;&lt;/i&gt; 11:24:02  

&lt;b&gt;2022&lt;/b&gt; • [SciLifeLab](https://www.scilifelab.se/) • [NBIS](https://nbis.se/) • [RaukR](https://nbisweden.github.io/workshop-RaukR-2206/)


    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="assets/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "4:3",
"highlightLanguage": "r",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "RaukR 2022 • %current%/%total%"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
